File: spark-scala-maven/src/main/scala/adj/spark/df/RDD2DataFrameProgrammatically.scala
package adj.spark.df

import org.apache.spark.sql.SQLContext
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import adj.spark.df.pojo.Student
import com.alibaba.fastjson.JSON
import org.apache.spark.sql.RowFactory
import scala.collection.immutable.List
import org.apache.spark.sql.types.DataTypes
import org.apache.spark.sql.types.StructType

object RDD2DataFrameProgrammatically {
  def main(args: Array[String]): Unit = {
     val conf = new SparkConf()
                    .setAppName(RDD2DataFrameProgrammatically.getClass.getSimpleName)
                    .setMaster("local")
    val context     = new SparkContext(conf)
    val sqlContext  = new SQLContext(context)
    
    val lineRDD = context.textFile("hdfs://master:9000/user/spark/data/students.json", 3)
    
    val studentRDD = lineRDD.map(line => {
      val js   = JSON.parseObject(line)
      
      val id   = js.getInteger("id")
      val name = js.getString("name")
      val age  = js.getInteger("age")
      
      RowFactory.create(id, name, age)
    })
    
    // 编程方式动态构造元数据
    val fieldTypes = List(DataTypes.createStructField("id", DataTypes.IntegerType, true),
                          DataTypes.createStructField("name", DataTypes.StringType, true),
                          DataTypes.createStructField("age", DataTypes.IntegerType, true))
    
    // RDD转换为DataFrame
    val structType = StructType(fieldTypes)
    
    val studentDF = sqlContext.createDataFrame(studentRDD, structType)
                          
    studentDF.registerTempTable("students")
    
    /**
     * 调用SqlContext.sql方法, 用SqlParser生成一个Unresolved LogicalPlan,
     * 并且封装在DataFrame中。
     */
    val sqlStudentDF = sqlContext.sql("select id, name, age from students")
    /**
     * 调用DataFrame打印操作,或者rdd的transaformation、action操作,就会触发SQLContext
     * 的executeSql操作,executeSql调用执行QueryExecution的操作。
     * 
     * 然后,调用Analyzer的apply方法,使用Unresolved LogicalPlan生成Resolved LogicalPlan,
     * 也就是将LogicalPlan与SQL语句的数据源执行绑定。
     */
    sqlStudentDF.show()
    
    val studentResultRDD = sqlStudentDF.rdd.map(row => {
      val id   = row.getAs[Int]("id")
      val name = row.getAs[String]("name")
      val age  = row.getInt(2)
      
      Student(id, name, age)
    })
    
    val studentResultRDD2 = sqlStudentDF.rdd.map(row => {
      val rowMap = row.getValuesMap[Any](Array("id", "name", "age"))
      val id   = rowMap("id").toString().toInt
      val name = "_" + rowMap("name").toString()
      val age  = rowMap("age").toString().toInt
      
      Student(id, name, age)
    })
    
    studentResultRDD.foreach(println)
    studentResultRDD2.foreach(println)
  }
}

File: spark-java/src/main/java/adj/spark/df/RDD2DataFrameProgrammatically.java
package adj.spark.df;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import com.alibaba.fastjson.JSONObject;

import adj.spark.df.pojo.Student;

/**
 * 使用编程方式动态指定元数据, 将RDD转换为DataFrame
 * 
 * @author felix
 */
public class RDD2DataFrameProgrammatically {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf()
				         .setMaster("local")
				         .setAppName("RDD2DataFrameReflection");

		JavaSparkContext context    = new JavaSparkContext(conf);
		SQLContext       sqlContext = new SQLContext(context);

		JavaRDD<String>  lineRDD = context.textFile("hdfs://master:9000/user/spark/data/students.json");
		JavaRDD<Row>     rowRDD  = lineRDD.map(new Function<String, Row>() {
			private static final long serialVersionUID = 1L;

			@Override
			public Row call(String line) throws Exception {
				JSONObject json = JSONObject.parseObject(line);
				int    id   = json.getInteger("id");
				String name = json.getString("name");
				int    age  = json.getInteger("age");

				return RowFactory.create(id, name, age);
			}
		});

		/**
		 * 动态构造元数据
		 * 适合于从数据库加载数据, 动态创建
		 */
		List<StructField> fieldTypes = new ArrayList<>();
		fieldTypes.add(DataTypes.createStructField("id", DataTypes.IntegerType, true));
		fieldTypes.add(DataTypes.createStructField("name", DataTypes.StringType, true));
		fieldTypes.add(DataTypes.createStructField("age", DataTypes.IntegerType, true));
		
		StructType structType = DataTypes.createStructType(fieldTypes);
		
		// 使用动态构造的元数据,将RDD转换为DataFrame
		DataFrame studentDataFrame = sqlContext.createDataFrame(rowRDD, structType);

		// 拿到一个DataFrame之后,就可以将其注册为一个临时表,然后针对其中的数据执行SQL语句
		studentDataFrame.registerTempTable("students");

		/**
		 * 针对students临时表,执行SQL语句。 注意：如果没有指明查询字段,查询的结果是乱序的
		 */
		DataFrame sqlDataFrame = sqlContext.sql("select id,name,age from students where age < 19");

		// 显示执行的结果
		sqlDataFrame.show();

		// 将DataFrame转换为RDD
		JavaRDD<Row> sqlRowRDD = sqlDataFrame.javaRDD();

		JavaRDD<Student> sqlStudentRDD = sqlRowRDD.map(new Function<Row, Student>() {
			private static final long serialVersionUID = 1L;

			@Override
			public Student call(Row row) throws Exception {
				int id = row.getInt(0);
				String name = "_" + row.getString(1);
				int age = row.getInt(2);
				return new Student(id, name, age);
			}
		});

		sqlStudentRDD.foreach(new VoidFunction<Student>() {
			private static final long serialVersionUID = 1L;

			@Override
			public void call(Student t) throws Exception {
				System.out.println(t);
			}
		});
	}
}
